// molstar/lib/commonjs/mol-plugin/behavior/dynamic/camera.js
"use strict";
/**
 * Copyright (c) 2018-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author Jason Pattle <jpattle.exscientia.co.uk>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CameraControls = exports.CameraAxisHelper = exports.FocusLoci = exports.DefaultFocusLociBindings = exports.DefaultClickResetCameraOnEmptySelectMode = exports.DefaultClickResetCameraOnEmpty = void 0;
var tslib_1 = require("tslib");
var loci_1 = require("../../../mol-model/loci");
var param_definition_1 = require("../../../mol-util/param-definition");
var behavior_1 = require("../behavior");
var input_observer_1 = require("../../../mol-util/input/input-observer");
var binding_1 = require("../../../mol-util/binding");
var commands_1 = require("../../commands");
var camera_helper_1 = require("../../../mol-canvas3d/helper/camera-helper");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var B = input_observer_1.ButtonsType;
var M = input_observer_1.ModifiersKeys;
var Trigger = binding_1.Binding.Trigger;
var Key = binding_1.Binding.TriggerKey;
exports.DefaultClickResetCameraOnEmpty = (0, binding_1.Binding)([
    Trigger(B.Flag.Primary, M.create()),
    Trigger(B.Flag.Secondary, M.create()),
    Trigger(B.Flag.Primary, M.create({ control: true }))
], 'Reset camera focus', 'Click on nothing using ${triggers}');
exports.DefaultClickResetCameraOnEmptySelectMode = (0, binding_1.Binding)([
    Trigger(B.Flag.Secondary, M.create()),
    Trigger(B.Flag.Primary, M.create({ control: true }))
], 'Reset camera focus', 'Click on nothing using ${triggers}');
exports.DefaultFocusLociBindings = {
    clickCenterFocus: (0, binding_1.Binding)([
        Trigger(B.Flag.Primary, M.create()),
        Trigger(B.Flag.Secondary, M.create()),
        Trigger(B.Flag.Primary, M.create({ control: true }))
    ], 'Camera center and focus', 'Click element using ${triggers}'),
    clickCenterFocusSelectMode: (0, binding_1.Binding)([
        Trigger(B.Flag.Secondary, M.create()),
        Trigger(B.Flag.Primary, M.create({ control: true }))
    ], 'Camera center and focus', 'Click element using ${triggers}'),
    clickResetCameraOnEmpty: exports.DefaultClickResetCameraOnEmpty,
    clickResetCameraOnEmptySelectMode: exports.DefaultClickResetCameraOnEmptySelectMode,
};
var FocusLociParams = {
    minRadius: param_definition_1.ParamDefinition.Numeric(8, { min: 1, max: 50, step: 1 }),
    extraRadius: param_definition_1.ParamDefinition.Numeric(4, { min: 1, max: 50, step: 1 }, { description: 'Value added to the bounding-sphere radius of the Loci' }),
    durationMs: param_definition_1.ParamDefinition.Numeric(250, { min: 0, max: 1000, step: 1 }, { description: 'Camera transition duration' }),
    bindings: param_definition_1.ParamDefinition.Value(exports.DefaultFocusLociBindings, { isHidden: true }),
};
exports.FocusLoci = behavior_1.PluginBehavior.create({
    name: 'camera-focus-loci',
    category: 'interaction',
    ctor: /** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.register = function () {
            var _this = this;
            this.subscribeObservable(this.ctx.behaviors.interaction.click, function (_a) {
                var _b, _c;
                var current = _a.current, button = _a.button, modifiers = _a.modifiers;
                if (!_this.ctx.canvas3d)
                    return;
                var binding = _this.ctx.selectionMode
                    ? _this.params.bindings.clickCenterFocusSelectMode
                    : _this.params.bindings.clickCenterFocus;
                var resetBinding = _this.ctx.selectionMode
                    ? ((_b = _this.params.bindings.clickResetCameraOnEmptySelectMode) !== null && _b !== void 0 ? _b : exports.DefaultClickResetCameraOnEmptySelectMode)
                    : ((_c = _this.params.bindings.clickResetCameraOnEmpty) !== null && _c !== void 0 ? _c : exports.DefaultClickResetCameraOnEmpty);
                if (loci_1.Loci.isEmpty(current.loci) && binding_1.Binding.match(resetBinding, button, modifiers)) {
                    commands_1.PluginCommands.Camera.Reset(_this.ctx, {});
                    return;
                }
                if (binding_1.Binding.match(binding, button, modifiers)) {
                    var loci = loci_1.Loci.normalize(current.loci, _this.ctx.managers.interactivity.props.granularity);
                    _this.ctx.managers.camera.focusLoci(loci, _this.params);
                }
            });
        };
        return class_1;
    }(behavior_1.PluginBehavior.Handler)),
    params: function () { return FocusLociParams; },
    display: { name: 'Camera Focus Loci on Canvas' }
});
exports.CameraAxisHelper = behavior_1.PluginBehavior.create({
    name: 'camera-axis-helper',
    category: 'interaction',
    ctor: /** @class */ (function (_super) {
        tslib_1.__extends(class_2, _super);
        function class_2() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_2.prototype.register = function () {
            var _this = this;
            var lastPlane = camera_helper_1.CameraHelperAxis.None;
            var state = 0;
            this.subscribeObservable(this.ctx.behaviors.interaction.click, function (_a) {
                var current = _a.current;
                if (!_this.ctx.canvas3d || !(0, camera_helper_1.isCameraAxesLoci)(current.loci))
                    return;
                var axis = current.loci.elements[0].groupId;
                if (axis === camera_helper_1.CameraHelperAxis.None) {
                    lastPlane = camera_helper_1.CameraHelperAxis.None;
                    state = 0;
                    return;
                }
                var camera = _this.ctx.canvas3d.camera;
                var dir, up;
                if (axis >= camera_helper_1.CameraHelperAxis.X && axis <= camera_helper_1.CameraHelperAxis.Z) {
                    lastPlane = camera_helper_1.CameraHelperAxis.None;
                    state = 0;
                    var d = linear_algebra_1.Vec3.sub((0, linear_algebra_1.Vec3)(), camera.target, camera.position);
                    var c = linear_algebra_1.Vec3.cross((0, linear_algebra_1.Vec3)(), d, camera.up);
                    up = (0, linear_algebra_1.Vec3)();
                    up[axis - 1] = 1;
                    dir = linear_algebra_1.Vec3.cross((0, linear_algebra_1.Vec3)(), up, c);
                    if (linear_algebra_1.Vec3.magnitude(dir) === 0)
                        dir = d;
                }
                else {
                    if (lastPlane === axis) {
                        state = (state + 1) % 2;
                    }
                    else {
                        lastPlane = axis;
                        state = 0;
                    }
                    if (axis === camera_helper_1.CameraHelperAxis.XY) {
                        up = state ? linear_algebra_1.Vec3.unitX : linear_algebra_1.Vec3.unitY;
                        dir = linear_algebra_1.Vec3.negUnitZ;
                    }
                    else if (axis === camera_helper_1.CameraHelperAxis.XZ) {
                        up = state ? linear_algebra_1.Vec3.unitX : linear_algebra_1.Vec3.unitZ;
                        dir = linear_algebra_1.Vec3.negUnitY;
                    }
                    else {
                        up = state ? linear_algebra_1.Vec3.unitY : linear_algebra_1.Vec3.unitZ;
                        dir = linear_algebra_1.Vec3.negUnitX;
                    }
                }
                _this.ctx.canvas3d.requestCameraReset({
                    snapshot: function (scene, camera) { return camera.getInvariantFocus(scene.boundingSphereVisible.center, scene.boundingSphereVisible.radius, up, dir); }
                });
            });
        };
        return class_2;
    }(behavior_1.PluginBehavior.Handler)),
    params: function () { return ({}); },
    display: { name: 'Camera Axis Helper' }
});
var DefaultCameraControlsBindings = {
    keySpinAnimation: (0, binding_1.Binding)([Key('KeyI')], 'Spin Animation', 'Press ${triggers}'),
    keyRockAnimation: (0, binding_1.Binding)([Key('KeyO')], 'Rock Animation', 'Press ${triggers}'),
    keyToggleFlyMode: (0, binding_1.Binding)([Key('Space', M.create({ shift: true }))], 'Toggle Fly Mode', 'Press ${triggers}'),
    keyResetView: (0, binding_1.Binding)([Key('KeyT')], 'Reset View', 'Press ${triggers}'),
};
var CameraControlsParams = {
    bindings: param_definition_1.ParamDefinition.Value(DefaultCameraControlsBindings, { isHidden: true }),
};
exports.CameraControls = behavior_1.PluginBehavior.create({
    name: 'camera-controls',
    category: 'interaction',
    ctor: /** @class */ (function (_super) {
        tslib_1.__extends(class_3, _super);
        function class_3() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_3.prototype.register = function () {
            var _this = this;
            this.subscribeObservable(this.ctx.behaviors.interaction.key, function (_a) {
                var code = _a.code, modifiers = _a.modifiers;
                if (!_this.ctx.canvas3d)
                    return;
                // include defaults for backwards state compatibility
                var b = tslib_1.__assign(tslib_1.__assign({}, DefaultCameraControlsBindings), _this.params.bindings);
                var p = _this.ctx.canvas3d.props.trackball;
                if (binding_1.Binding.matchKey(b.keySpinAnimation, code, modifiers)) {
                    var name_1 = p.animate.name !== 'spin' ? 'spin' : 'off';
                    if (name_1 === 'off') {
                        _this.ctx.canvas3d.setProps({
                            trackball: { animate: { name: name_1, params: {} } }
                        });
                    }
                    else {
                        _this.ctx.canvas3d.setProps({
                            trackball: { animate: {
                                    name: name_1,
                                    params: { speed: 1 }
                                }
                            }
                        });
                    }
                }
                if (binding_1.Binding.matchKey(b.keyRockAnimation, code, modifiers)) {
                    var name_2 = p.animate.name !== 'rock' ? 'rock' : 'off';
                    if (name_2 === 'off') {
                        _this.ctx.canvas3d.setProps({
                            trackball: { animate: { name: name_2, params: {} } }
                        });
                    }
                    else {
                        _this.ctx.canvas3d.setProps({
                            trackball: { animate: {
                                    name: name_2,
                                    params: { speed: 0.3, angle: 10 }
                                }
                            }
                        });
                    }
                }
                if (binding_1.Binding.matchKey(b.keyToggleFlyMode, code, modifiers)) {
                    var flyMode = !p.flyMode;
                    _this.ctx.canvas3d.setProps({
                        trackball: { flyMode: flyMode }
                    });
                    if (_this.ctx.canvas3dContext) {
                        _this.ctx.canvas3dContext.canvas.style.cursor = flyMode ? 'crosshair' : 'unset';
                    }
                }
                if (binding_1.Binding.matchKey(b.keyResetView, code, modifiers)) {
                    commands_1.PluginCommands.Camera.Reset(_this.ctx, {});
                }
            });
        };
        return class_3;
    }(behavior_1.PluginBehavior.Handler)),
    params: function () { return CameraControlsParams; },
    display: { name: 'Camera Controls on Canvas' }
});
