// molstar/lib/commonjs/mol-model-formats/structure/top.js
"use strict";
/**
 * Copyright (c) 2022 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.topologyFromTop = exports.TopFormat = void 0;
var tslib_1 = require("tslib");
var db_1 = require("../../mol-data/db");
var types_1 = require("../../mol-model/structure/model/types");
var topology_1 = require("../../mol-model/structure/topology/topology");
var mol_task_1 = require("../../mol-task");
var schema_1 = require("./basic/schema");
var component_1 = require("./common/component");
var entity_1 = require("./common/entity");
var util_1 = require("./common/util");
var util_2 = require("./util");
function getBasic(top) {
    var molecules = top.molecules, compounds = top.compounds;
    var singleResidue = {};
    var atomCount = 0;
    for (var i = 0, il = molecules._rowCount; i < il; ++i) {
        var mol = molecules.compound.value(i);
        var count = molecules.molCount.value(i);
        var atoms = compounds[mol].atoms;
        db_1.Column.asArrayColumn(atoms.atom);
        db_1.Column.asArrayColumn(atoms.resnr);
        db_1.Column.asArrayColumn(atoms.residu);
        atomCount += count * atoms._rowCount;
        var prevResnr = atoms.resnr.value(0);
        singleResidue[mol] = true;
        for (var j = 1, jl = atoms._rowCount; j < jl; ++j) {
            var resnr = atoms.resnr.value(j);
            if (resnr !== prevResnr) {
                singleResidue[mol] = false;
                break;
            }
            prevResnr = resnr;
        }
    }
    //
    var atomNames = new Array(atomCount);
    var residueIds = new Uint32Array(atomCount);
    var residueNames = new Array(atomCount);
    var k = 0;
    for (var i = 0, il = molecules._rowCount; i < il; ++i) {
        var mol = molecules.compound.value(i);
        var count = molecules.molCount.value(i);
        var atoms = compounds[mol].atoms;
        var isSingleResidue = singleResidue[mol];
        for (var j = 0; j < count; ++j) {
            for (var l = 0, ll = atoms._rowCount; l < ll; ++l) {
                atomNames[k] = atoms.atom.value(l);
                residueIds[k] = atoms.resnr.value(l);
                residueNames[k] = atoms.residu.value(l);
                if (isSingleResidue)
                    residueIds[k] += j;
                k += 1;
            }
        }
    }
    var atomName = db_1.Column.ofStringArray(atomNames);
    var residueId = db_1.Column.ofIntArray(residueIds);
    var residueName = db_1.Column.ofStringArray(residueNames);
    //
    var entityIds = new Array(atomCount);
    var asymIds = new Array(atomCount);
    var seqIds = new Uint32Array(atomCount);
    var ids = new Uint32Array(atomCount);
    var entityBuilder = new entity_1.EntityBuilder();
    var componentBuilder = new component_1.ComponentBuilder(residueId, atomName);
    var currentEntityId = '';
    var currentAsymIndex = 0;
    var currentAsymId = '';
    var currentSeqId = 0;
    var prevMoleculeType = 0 /* MoleculeType.Unknown */;
    var prevResidueNumber = -1;
    for (var i = 0, il = atomCount; i < il; ++i) {
        var residueNumber = residueId.value(i);
        if (residueNumber !== prevResidueNumber) {
            var compId = residueName.value(i);
            var moleculeType = (0, types_1.getMoleculeType)(componentBuilder.add(compId, i).type, compId);
            if (moleculeType !== prevMoleculeType) {
                currentAsymId = (0, util_1.getChainId)(currentAsymIndex);
                currentAsymIndex += 1;
                currentSeqId = 0;
            }
            currentEntityId = entityBuilder.getEntityId(compId, moleculeType, currentAsymId);
            currentSeqId += 1;
            prevResidueNumber = residueNumber;
            prevMoleculeType = moleculeType;
        }
        entityIds[i] = currentEntityId;
        asymIds[i] = currentAsymId;
        seqIds[i] = currentSeqId;
        ids[i] = i;
    }
    var id = db_1.Column.ofIntArray(ids);
    var asym_id = db_1.Column.ofStringArray(asymIds);
    //
    var type_symbol = new Array(atomCount);
    for (var i = 0; i < atomCount; ++i) {
        type_symbol[i] = (0, util_2.guessElementSymbolString)(atomName.value(i), residueName.value(i));
    }
    var atom_site = db_1.Table.ofPartialColumns(schema_1.BasicSchema.atom_site, {
        auth_asym_id: asym_id,
        auth_atom_id: db_1.Column.asArrayColumn(atomName),
        auth_comp_id: residueName,
        auth_seq_id: residueId,
        id: db_1.Column.asArrayColumn(id),
        label_asym_id: asym_id,
        label_atom_id: db_1.Column.asArrayColumn(atomName),
        label_comp_id: residueName,
        label_seq_id: db_1.Column.ofIntArray(seqIds),
        label_entity_id: db_1.Column.ofStringArray(entityIds),
        occupancy: db_1.Column.ofConst(1, atomCount, db_1.Column.Schema.float),
        type_symbol: db_1.Column.ofStringArray(type_symbol),
        pdbx_PDB_model_num: db_1.Column.ofConst(1, atomCount, db_1.Column.Schema.int),
    }, atomCount);
    var basic = (0, schema_1.createBasic)({
        entity: entityBuilder.getEntityTable(),
        chem_comp: componentBuilder.getChemCompTable(),
        atom_site: atom_site
    });
    return basic;
}
function getBonds(top) {
    var molecules = top.molecules, compounds = top.compounds;
    var indexA = [];
    var indexB = [];
    var atomOffset = 0;
    for (var i = 0, il = molecules._rowCount; i < il; ++i) {
        var mol = molecules.compound.value(i);
        var count = molecules.molCount.value(i);
        var _a = compounds[mol], atoms = _a.atoms, bonds = _a.bonds;
        if (bonds) {
            for (var j = 0; j < count; ++j) {
                for (var l = 0, ll = bonds._rowCount; l < ll; ++l) {
                    indexA.push(bonds.ai.value(l) - 1 + atomOffset);
                    indexB.push(bonds.aj.value(l) - 1 + atomOffset);
                }
                atomOffset += atoms._rowCount;
            }
        }
        else if (mol === 'TIP3') {
            for (var j = 0; j < count; ++j) {
                indexA.push(0 + atomOffset);
                indexB.push(1 + atomOffset);
                indexA.push(0 + atomOffset);
                indexB.push(2 + atomOffset);
                atomOffset += atoms._rowCount;
            }
        }
        else {
            atomOffset += count * atoms._rowCount;
        }
    }
    return {
        indexA: db_1.Column.ofIntArray(indexA),
        indexB: db_1.Column.ofIntArray(indexB),
        order: db_1.Column.ofConst(1, indexA.length, db_1.Column.Schema.int)
    };
}
var TopFormat;
(function (TopFormat) {
    function is(x) {
        return (x === null || x === void 0 ? void 0 : x.kind) === 'top';
    }
    TopFormat.is = is;
    function fromTop(top) {
        return { kind: 'top', name: top.system || 'TOP', data: top };
    }
    TopFormat.fromTop = fromTop;
})(TopFormat || (TopFormat = {}));
exports.TopFormat = TopFormat;
function topologyFromTop(top) {
    var _this = this;
    return mol_task_1.Task.create('Parse TOP', function (ctx) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var format, basic, bonds;
        return tslib_1.__generator(this, function (_a) {
            format = TopFormat.fromTop(top);
            basic = getBasic(top);
            bonds = getBonds(top);
            return [2 /*return*/, topology_1.Topology.create(top.system || 'TOP', basic, bonds, format)];
        });
    }); });
}
exports.topologyFromTop = topologyFromTop;
